File: app/src/main/java/com/spbstu/sneakerhunter/fragments/FiltersFragment.java
package com.spbstu.sneakerhunter.fragments;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;

import com.spbstu.sneakerhunter.R;
import com.spbstu.sneakerhunter.server_list.Brand;
import com.spbstu.sneakerhunter.server_list.Size;
import com.spbstu.sneakerhunter.server_list.Sneaker;
import com.spbstu.sneakerhunter.server_list.SneakersAPI;

import java.util.List;
import java.util.stream.Collectors;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class FiltersFragment extends Fragment {

    private Retrofit retrofit;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    private Retrofit getClient() {
        if(retrofit == null){
            retrofit = new Retrofit.Builder()
                    .baseUrl(SneakersAPI.URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }


    public void createSpinner(View view) {

        SneakersAPI sneakersAPI = getClient().create(SneakersAPI.class);

        Call<List<Size>> sizesCall = sneakersAPI.getSizes();
        sizesCall.enqueue(new Callback<List<Size>>() {
            @Override
            public void onResponse(Call<List<Size>> call, Response<List<Size>> response) {
                if (response.isSuccessful()) {
                    List<Size> sizes = response.body().stream().sorted().collect(Collectors.toList());

                    String[] sizesStrings = new String[sizes.size()+1];
                    sizesStrings[0] = "Не выбрано";
                    for (int i = 0; i < sizes.size(); i++)
                        sizesStrings[i+1] = sizes.get(i).getSize();

                    Spinner spinner = view.findViewById(R.id.sizeSpinner);
                    ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(),
                            android.R.layout.simple_spinner_dropdown_item, sizesStrings);
                    spinner.setAdapter(adapter);
                }
            }

            @Override
            public void onFailure(Call<List<Size>> call, Throwable t) {
                System.out.println("fail: " + t);
            }
        });
    }

    public void createBrandsSpinner(View view) {

        SneakersAPI sneakersAPI = getClient().create(SneakersAPI.class);

        Call<List<Brand>> sizesCall = sneakersAPI.getBrands();
        sizesCall.enqueue(new Callback<List<Brand>>() {
            @Override
            public void onResponse(Call<List<Brand>> call, Response<List<Brand>> response) {
                if (response.isSuccessful()) {
                    List<Brand> brands = response.body().stream().sorted().collect(Collectors.toList());

                    String[] sizesStrings = new String[brands.size()+1];
                    sizesStrings[0] = "Не выбрано";
                    for (int i = 0; i < brands.size(); i++)
                        sizesStrings[i+1] = brands.get(i).getName();

                    Spinner spinner = view.findViewById(R.id.brandsSpinner);
                    ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(),
                            android.R.layout.simple_spinner_dropdown_item, sizesStrings);
                    spinner.setAdapter(adapter);
                }
            }

            @Override
            public void onFailure(Call<List<Brand>> call, Throwable t) {
                System.out.println("fail: " + t);
            }
        });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_filters, container, false);

        Toolbar toolbar = (Toolbar) view.findViewById(R.id.toolbar);
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
        ((AppCompatActivity) getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        ((AppCompatActivity) getActivity()).getSupportActionBar().setDisplayShowTitleEnabled(false);

        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getActivity().getSupportFragmentManager().popBackStack();
            }
        });

        createSpinner(view);
        createBrandsSpinner(view);

        Spinner shopSpin = (Spinner) view.findViewById(R.id.brandSpinner);
        ArrayAdapter<String> adapterShop = new ArrayAdapter<>(getContext(),
                android.R.layout.simple_spinner_dropdown_item, new String[]{"Не выбрано", "Asos", "Ali Express"});
        shopSpin.setAdapter(adapterShop);

        Spinner sizeSpinner  = (Spinner) view.findViewById(R.id.sizeSpinner);
        EditText priceFromText = (EditText) view.findViewById(R.id.priceFromEditText);
        EditText priceToText = (EditText) view.findViewById(R.id.priceToEditText);
        Spinner brandSpinner = (Spinner) view.findViewById(R.id.brandsSpinner);

        Button confirmButton = (Button) view.findViewById(R.id.confirm_button);
        confirmButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (!sizeSpinner.getSelectedItem().toString().equals("Не выбрано")) {
                    SearchFragment.SORT_SIZE = sizeSpinner.getSelectedItem().toString();
                    SearchFragment.IS_SORT_SIZE = true;
                }

                if(!priceFromText.getText().toString().equals("")) {
                    SearchFragment.SORT_PRICE_FROM = Double.parseDouble(priceFromText.getText().toString());
                    SearchFragment.IS_SORT_PRICE = true;
                }

                if(!priceToText.getText().toString().equals("")) {
                    SearchFragment.SORT_PRICE_TO = Double.parseDouble(priceToText.getText().toString());
                    SearchFragment.IS_SORT_PRICE = true;
                }

                if (!shopSpin.getSelectedItem().toString().equals("Не выбрано")) {
                    SearchFragment.SORT_SHOP = shopSpin.getSelectedItem().toString();
                    SearchFragment.IS_SORT_SHOP = true;
                }

                if (!brandSpinner.getSelectedItem().toString().equals("Не выбрано")) {
                    SearchFragment.SORT_BRAND = brandSpinner.getSelectedItem().toString();
                    SearchFragment.IS_SORT_BRAND = true;
                }

                getActivity().getSupportFragmentManager().popBackStack();
            }
        });

        Button resetButton = (Button) view.findViewById(R.id.reset_button);
        resetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SearchFragment.IS_SORT_SIZE = false;
                SearchFragment.IS_SORT_PRICE = false;
                SearchFragment.IS_SORT_SHOP = false;
                SearchFragment.IS_SORT_BRAND = false;

                sizeSpinner.setSelection(0);
                shopSpin.setSelection(0);
                priceFromText.setText("");
                priceToText.setText("");

                getActivity().getSupportFragmentManager().popBackStack();
            }
        });

        return view;
    }
}

File: app/src/main/java/com/spbstu/sneakerhunterkotlin/fragments/FiltersFragment.kt
package com.spbstu.sneakerhunterkotlin.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.Spinner
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import com.spbstu.sneakerhunterkotlin.R
import com.spbstu.sneakerhunterkotlin.server_list.Brand
import com.spbstu.sneakerhunterkotlin.server_list.Size
import com.spbstu.sneakerhunterkotlin.server_list.SneakersAPI
import io.ktor.utils.io.core.String
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.stream.Collectors

class FiltersFragment : Fragment() {
    private lateinit var retrofit: Retrofit
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    private val client: Retrofit
        get() {
            retrofit = Retrofit.Builder()
                .baseUrl(SneakersAPI.URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
            return retrofit
        }

    private fun createSpinner(view: View) {
        val sneakersAPI: SneakersAPI = client.create(SneakersAPI::class.java)
        val sizesCall: Call<List<Size?>?>? = sneakersAPI.sizes

        sizesCall?.enqueue(object : Callback<List<Size?>?> {
            override fun onResponse(call: Call<List<Size?>?>, response: Response<List<Size?>?>) {
                if (response.isSuccessful) {
                    val sizes: MutableList<Size?>? = response.body()!!.stream().sorted()
                        .collect(Collectors.toList())
                    val sizesStrings =
                        arrayOfNulls<String>(sizes!!.size + 1)
                    sizesStrings[0] = "Не выбрано"
                    for (i in sizes.indices) sizesStrings[i + 1] = sizes[i]?.size
                    val spinner = view.findViewById<Spinner>(R.id.sizeSpinner)
                    val adapter = ArrayAdapter(
                        context!!,
                        android.R.layout.simple_spinner_dropdown_item, sizesStrings
                    )
                    spinner.adapter = adapter
                    for (i in 0 until spinner.count) {
                        if (spinner.getItemAtPosition(i)
                                .toString() == SearchFragment.SORT_SIZE
                        ) spinner.setSelection(i)
                    }
                }
            }

            override fun onFailure(call: Call<List<Size?>?>, t: Throwable) {
                println("fail: $t")
            }
        })
    }

    private fun createBrandsSpinner(view: View) {
        val sneakersAPI: SneakersAPI = client.create(SneakersAPI::class.java)
        val sizesCall: Call<List<Brand?>?>? = sneakersAPI.brands

        sizesCall?.enqueue(object : Callback<List<Brand?>?> {
            override fun onResponse(call: Call<List<Brand?>?>, response: Response<List<Brand?>?>) {
                if (response.isSuccessful) {
                    val brands: MutableList<Brand?>? = response.body()!!.stream().sorted()
                        .collect(Collectors.toList())
                    val sizesStrings =
                        arrayOfNulls<String>(brands!!.size + 1)
                    sizesStrings[0] = "Не выбрано"
                    for (i in brands.indices) sizesStrings[i + 1] = brands[i]?.name
                    val spinner = view.findViewById<Spinner>(R.id.brandsSpinner)
                    val adapter = ArrayAdapter(
                        context!!,
                        android.R.layout.simple_spinner_dropdown_item, sizesStrings
                    )
                    spinner.adapter = adapter
                    for (i in 0 until spinner.count) {
                        if (spinner.getItemAtPosition(i)
                                .toString() == SearchFragment.SORT_BRAND
                        ) spinner.setSelection(i)
                    }
                }
            }

            override fun onFailure(call: Call<List<Brand?>?>, t: Throwable) {
                println("fail: $t")
            }
        })
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view: View = inflater.inflate(R.layout.fragment_filters, container, false)
        val toolbar =
            view.findViewById<View>(R.id.toolbar) as Toolbar
        (activity as AppCompatActivity).setSupportActionBar(toolbar)
        (activity as AppCompatActivity).supportActionBar?.setDisplayHomeAsUpEnabled(true)
        (activity as AppCompatActivity).supportActionBar?.setDisplayShowTitleEnabled(false)
        toolbar.setNavigationOnClickListener {
            requireActivity().supportFragmentManager.popBackStack()
        }

        createSpinner(view)
        createBrandsSpinner(view)
        val shopSpin = view.findViewById<View>(R.id.brandSpinner) as Spinner
        val adapterShop = ArrayAdapter(
            requireContext(),
            android.R.layout.simple_spinner_dropdown_item,
            arrayOf("Не выбрано", "Asos", "Ali Express")
        )

        shopSpin.adapter = adapterShop
        for (i in 0 until shopSpin.count) {
            if (shopSpin.getItemAtPosition(i).toString() == SearchFragment.SORT_SHOP)
                shopSpin.setSelection(i)
        }

        val sizeSpinner = view.findViewById<View>(R.id.sizeSpinner) as Spinner
        val priceFromText = view.findViewById<View>(R.id.priceFromEditText) as EditText

        val priceToText = view.findViewById<View>(R.id.priceToEditText) as EditText

        val brandSpinner = view.findViewById<View>(R.id.brandsSpinner) as Spinner

        val confirmButton = view.findViewById<View>(R.id.confirm_button) as Button
        confirmButton.setOnClickListener {
            if (sizeSpinner.selectedItem.toString() != "Не выбрано") {
                SearchFragment.SORT_SIZE = sizeSpinner.selectedItem.toString()
                SearchFragment.IS_SORT_SIZE = true
            }
            if (priceFromText.text.toString() != "") {
                SearchFragment.SORT_PRICE_FROM = priceFromText.text.toString().toDouble()
                SearchFragment.IS_SORT_PRICE = true
            }
            if (priceToText.text.toString() != "") {
                SearchFragment.SORT_PRICE_TO = priceToText.text.toString().toDouble()
                SearchFragment.IS_SORT_PRICE = true
            }
            if (shopSpin.selectedItem.toString() != "Не выбрано") {
                SearchFragment.SORT_SHOP = shopSpin.selectedItem.toString()
                SearchFragment.IS_SORT_SHOP = true
            }
            if (brandSpinner.selectedItem.toString() != "Не выбрано") {
                SearchFragment.SORT_BRAND = brandSpinner.selectedItem.toString()
                SearchFragment.IS_SORT_BRAND = true
            }
            requireActivity().supportFragmentManager.popBackStack()
        }

        val resetButton = view.findViewById<View>(R.id.reset_button) as Button
        resetButton.setOnClickListener {
            SearchFragment.IS_SORT_SIZE = false
            SearchFragment.SORT_PRICE_FROM = 0.0
            SearchFragment.IS_SORT_PRICE = false
            SearchFragment.SORT_PRICE_TO = 0.0
            SearchFragment.IS_SORT_SHOP = false
            SearchFragment.IS_SORT_BRAND = false
            sizeSpinner.setSelection(0)
            shopSpin.setSelection(0)
            priceFromText.setText("")
            priceToText.setText("")
            requireActivity().supportFragmentManager.popBackStack()
        }
        return view
    }
}

